3.5 WIP: Hangprinter: Optionally allow up to 5 anchors/motors to be configured

jtimon opened the pull request:
This is a rebase of a subset of  https://github.com/Duet3D/RepRapFirmware/pull/585 which can be seen in https://github.com/GloomyAndy/RepRapFirmware/compare/v3.4-dev...jtimon:v3.4-hangprinter-5anchors?expand=1

This is not complete yet though, because I just passed by the flex stuff. I hope it doesn't take me much to "complete" this PR in that sense. 

The 5 anchor version must be the simplest to optionally support. 6 should be easy too, I think.
In 3.4, the first problems appeared with 7 anchors or more, because the config messages didn't fit. which I got some guidance for here: https://github.com/Duet3D/RepRapFirmware/pull/585#discussion_r973196192
But I didn't do it because 3.5 was out and I needed to rebase. 

Currently this shouldn't work for 5 anchors yet but it also shouldn't break 4 anchor hangprinters.
I didn't test anything yet for 3.5.

But it's ready for suggestion for changes from review or testing. 

This PR contains the following PRs:

- [ ]  3.5 NOCHANGE: hangprinter: simplify things by turning unrolled stuff into loops #597 
- [ ]  3.5 Hangprinter: Generalize IsReachable() to any pyramid #598 

TODO:

- [ ] Fully adapt to the changes from 3.4 to 3.5
- [ ] (kind of optional) Adapt HangprinterKinematics::MotorStepsToCartesian to N anchors for better accuracy.

jtimon committed:
Hangprinter: Generalize IsReachable() to any pseudo-pyramid

This is a clean generalization from 4 to N.
Pseudo-pyramid because the vertices in the base of the pseudo-pyramid
don't need to be on the same plane. There's no base of the pyramid as such.

Better documented and cleaner thanks to @tobbelobb on github.

jtimon committed:
Hangprinter: Loop around the anchors instead of naming them directly

jtimon committed:
WIP: Hangprinter: Allow Hangprinter to be configured with 5 anchors

This allows to print square pyramids instead of tetrahedrons, at the
cost of an extra motor.

WIP: not everything is adapted after recent changes. See the newest TODOs

jtimon committed:
HangprinterKinematics::WriteCalibrationParameters: flat return style

Otherwise it gets too deep.
I think there's an option in eclipse to forbid this by project in
format, one can config depness iirc.

tobbelobb commented:
Nice!

> // Precondition: all the base anchors belong to the same plane,

I don't think we should add this precondition. Better to have an imperfect `IsReachable()` implementation than to place restrictions on users.

tobbelobb commented:
I agree that this PR allows more configurations. But I think this PR is a good opportunity to remove the base plane check even though there was a base plane check in the previous version of `IsReachable()`. The base plane concept was awkward from before and trying to generalize it is even more awkward.

The base plane check was there just to save users from destroying hardware if they did a typo like `G1 Z-1000` or similar. Since I now understand that the `z_min` limit would save those users anyways I think we should drop the base plane check in `IsReachable()`.

>  I'm not really sure there will be much interest in supporting pseudo-pyramids with not all the anchors in the same plane, to be honest.

In practice all 1h4l-Hangprinters will be pseudo-pyramids with not all anchors in the same plane, since placing four anchors in a flat plane is hard to do (even getting within 10 mm is sometimes hard in my experience), and adds no value.

There is some interest in pseudo-pyramids and non-flat-bottomed pyramids, a few users including myself have used non-flat pyramids a few times already. One example use case is when I move anchors mid-print while printing very tall objects, I tend to move one anchor at a time to reduce risk, giving very non-horizontal intermediary ABC planes that must work for these super-tall objects to be printable.

But my main reason for not wanting the base plane concept (although I introduced it) is the effort it would take to explain users that ABC spans a plane, not BCD, CDA, or DAB, and that moving below this plane is disallowed in firmware (why?).

Sorry for asking you to remove the base plane check that I added long ago, but I didn't realize how misplaced it was before I tried to generalize it while I was reviewing this PR. If you prefer, I can wedge a commit where I remove the old base plane check before this commit, or I can remove it in a separate commit after this one.

tobbelobb commented:
> Anyway, if you have an alternative to https://github.com/Duet3D/RepRapFirmware/pull/598 that is more generic than this, all welcomed.

Oh, I have been commenting on the wrong PR. This one is a ok :ok_hand:
It's the other one that changes `IsReachable()`. Sry.

jtimon commented:
Well, it applies to this one too since this one contains the other one, but, yeah, I guess better discuss it there if it's specific to that.

jtimon committed:
HP: Allow custom and prism-like setups instead of just pyramid-like setups

- Change from bool to class and introduce None mode for
custom (experimental) setups (all allowed)

- refactor out IsInsidePyramidSides

- refactor out IsInsidePrismSides

- introduce HangprinterAnchorMode::LastOnTop for prism-like setups
  with all anchors on top

- introduce a T parameter in M666 to configure the anchor mode

tobbelobb commented on src/Movement/Kinematics/HangprinterKinematics.cpp:
```diff
@@ -342,32 +402,33 @@ inline float HangprinterKinematics::MotorPosToLinePos(const int32_t motorPos, si
 
 void HangprinterKinematics::flexDistances(float const machinePos[3], float const distanceA,
```
This function must take an array of numAnchors distances that must be looped over

jtimon replied:
Yes, adapt everything in #569 to not using A_AXIS constants inside.
But first deal with HangprinterKinematics::StaticForces and the other changes in #562

tobbelobb commented:
Only had my phone available so only a brief review. Noticed that flex compensation isn't yet rolled into loops over numAnchors. Need to finish that before we can test on hw

tobbelobb commented on src/Movement/Kinematics/HangprinterKinematics.cpp:
```diff
@@ -532,69 +607,115 @@ AxesBitmap HangprinterKinematics::MustBeHomedAxes(AxesBitmap axesMoving, bool di
+	scratchString.printf("M669 K6 ");
+	ok = f->Write(scratchString.c_str());
+	if (!ok) return false;
+
```
We should probably write numAnchors to the string around here

tobbelobb commented on src/Movement/Kinematics/HangprinterKinematics.h:
```diff
@@ -57,7 +57,8 @@ class HangprinterKinematics : public RoundBedKinematics
 private:
 	// Basic facts about movement system
-	static constexpr size_t HANGPRINTER_AXES = 4;
+	const char* ANCHOR_CHARS = "ABCDE";
```
I think we discussed anchor chars some other time. The letter here E is confusing here since the extruder axis is already called E

tobbelobb replied:
I managed to find the previous discussion https://github.com/Duet3D/RepRapFirmware/pull/585#discussion_r1055352961

I'm still in favor of naming the axes ABCDIJKLO. The naming is important since it will face users and will probably go into build manuals and instruction videos. Later firmware implementations will maybe have to imitate our choice.

jtimon replied:
Yes, not only we need to maintain the first ABCD for backwards compatibility with any previous hangprinter, but also we shouldn't conflict with any future notion of extra axes beyond XYZ.
I want to rotate the head because I've seen we can. Not arbitrarily, but 3 extra axes in some angle depnding on how do you build it mechanically. 
That could serve for some tool changing systems, a bit of rotation may be enough.
Also that could serve give stability to alltop configs, possibly allowing them to move faster.
I know, that's more fast forward calculations because now you need to know where the anchors are attached in the head to rotate it precisely. 
In any case, I think that decision was right and I just forgot about it. It's the E, that's for the extruder, even worse.
But thanks for the reminder of the order we decided for anchors: ABCDIJKLO.
That must be fixed, thanks.

jtimon replied:
I'm going to use ABCDIJKLO directly even though that's not needed for 5 anchors because I will likely forget again and the loops accessing ANCHOR_CHARS should be properly limited anyway.
We could add a diff line for every new anchor:

--- ANCHOR_CHARS = "ABCDIJ"
+++ ANCHOR_CHARS = "ABCDIJK"

but I think it's fine to do it all at once. If not, I can easily remove JKLO and re-add it again in latter commits as needed.

tobbelobb replied:
I think adding all the chars at once is good :+1:

tobbelobb commented on src/Movement/Kinematics/HangprinterKinematics.h:
```diff
@@ -57,7 +57,8 @@ class HangprinterKinematics : public RoundBedKinematics
-	static constexpr size_t HANGPRINTER_AXES = 4;
+	const char* ANCHOR_CHARS = "ABCDE";
+	static constexpr size_t HANGPRINTER_MAX_ANCHORS = 5;
 	static constexpr size_t A_AXIS = 0;
```
We should add K_AXIS etc here, alternatively remove A_AXIS etc if they are not used anywhere

jtimon replied:
I think we should delete all A_AXIS to D_AXIS constants.
That's the only way to make sure you're lopping to maxAnchors everywhere. But we're not ready yet in that sense like we were in 3.4. There are new uses of A_AXIS since https://github.com/Duet3D/RepRapFirmware/pull/562/files
It's alright, we can get there again. The thing is, newer changes unrelated to this will add whatever is around them. If we loop around ANCHOR_CHARS they will use that adapting it to whatever they're changing. If we keep using these constants, more new uses of them will keep appearing.
I was never ready to remove them in 3.4 though, they were still used in the fastforwards thing.
So until we loop in there we still need them.

tobbelobb replied:
There are only your pull request going into HangprinterKinematics.h and HangprinterKinematics.cpp for the forseeable future, so we should be able to get rid of A_AXIS to D_AXIS without new uses popping up from elsewhere in the meantime. Let's do it

jtimon commented:
> Only had my phone available so only a brief review. Noticed that flex compensation isn't yet rolled into loops over numAnchors. Need to finish that before we can test on hw

Yes, that's the kind of thing I mean by "Work In Progress" (WIP).
It means there are unsolved things To Do in the new comments of the PR.

But it's normal to have to rebase. When you were coding the flex stuff you had to focus on the flex stuff working and merging fine with the current branch at the time, which it did.

But if you do a PR on top of #597 but completing it in the parts related to flex which this doesn't account for, that would be awesome. But if not I will have to understand these new flex parameter at some point anyway.
With M669 and M66 there's a lot of parameters I still don't understand, which is alright as long as I don't break them.
But can I test myself if I broke them without understanding it myself?

In the case of the latest commit, I think yes, I think it's broken at the start part reserving memory dynamically.
The last commit breaks my setup, don't try in yours, try the previous commits. and on this last commit thank you for the review, keep it coming.

But when I break the wifi of my machine with the latest commit because I don't know what I'm doing with the memory...it gets me back to the times of opening the box for changing binaries in the SD or, worse, talkingto the printer by pronterface and an usb cable. Both mean opening the machine, sigh.

Just please nobody run this code on their machine unless you want to see an error and want to take out the sd card to reinstall the binaries, which is a pain in the ass.

Any tester interested in this PR should test the "previous" PR #603 which won't break your printer hopefully like this PR at this point.

tobbelobb commented:
> if you do a PR on top of https://github.com/Duet3D/RepRapFirmware/pull/597 but completing it in the parts related to flex which this doesn't account for, that would be awesome.

:+1:  Will look into it. Since it's not merged yet maybe I can push directly into your PR instead of opening a new one? #597

tobbelobb commented:
This PR can be safely closed, since ideas and code from this PR has already been merged in #606 and then generalized over again in #1230.
